### DISCUSSION TITLE
Be more permissive with document/head use

### DISCUSSION THREAD

**Discussion by TyMick** [vercel/next.js#19720] [Ideas]
Spitballing here. I had an idea (thanks to @willemmulder for bouncing ideas with me) for a way to make `document/head` less confusing, because it sounds like it's still generating friction (#9794, #13230) even after adding the [no-document-viewport-meta warning](https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/errors/no-document-viewport-meta.md).

## Background

Looks like the current policy governing what's allowed in `document/head` was first [announced](https://github.com/vercel/next.js/issues/6919#issuecomment-500471583) in June 2019 by @Timer in #6919:

> After careful review we've decided this is the intended behavior. Head elements that are expected to be deduped should be added using `next/head`.
> The reason this surfaced is because we started defaulting a more commonly used `<meta>` tag -- not that something is working unexpectedly.

Tangentially related, here's @lfades's explanation of the difference between `_document.js` and `_app.js` from May 2020 in #13408:

> `next/head` is dynamic, you can use it multiple times, update it, remove it, and your head tag will reflect those changes.
>
> `_document.js` is only rendered once, in the server, and for your entire page, this is the one that generates the initial `html` tag that will be used for your page and it's not a place to save any tags that rely on dynamic data, or that only apply for some pages. A good use case for this one is for example to add a custom favicon link, or google analytics, something that you know only needs to be added once and has to be always there.
>
> `_app.js` is similar in the way that it's also applied globally, but it does get rendered again every time you switch the current page, so it's also dynamic (which is never the case for `_document.js`)

## The struggle

This sounds like the reason why a lot of people want to put a `<meta name="viewport"` or `<meta charSet` in their `_document`, because they don't intend for those to ever change in their app—it's "something [they] know only needs to be added once and has to always be there."

Totally makes sense why the current policy was put in place when those sensible [defaults](https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/packages/next/next-server/lib/head.tsx#L11-L17) were added, since with `_document` only being rendered once, you can't take anything _out_ of `document/head` that conflicts with `next/head`. Here's the idea, though: **what if we just let `document/head` win all the conflicts and only warn users when they create the conflict themselves?**

## The idea

Is it possible, after `document/head` populates the initial HTML `<head>` but before `next/head` injects into the existing HTML's `<head>`, to pass a read-only list of `document/head` children to `next/head`'s [filter](https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/packages/next/next-server/lib/head.tsx#L114-L156)? That way, `next/head` could remove anything that conflicts with `document/head`, including the defaults, since the defaults live in `next/head`.

The process would look like this:

1. `document/head` adds any elements defined within it to the initial HTML, without concerning itself about defaults or anything. No warnings necessary.
2. The list of elements that `document/head` added gets passed to `next/head`'s filter somehow.
3. `next/head` filters out _all_ elements that conflict with `document/head` children.
   - If a conflict filtered out was part of `defaultHead`, that's perfect.
   - If a conflict filtered out was a manually defined `next/head` child, _then_ generate a warning, because the user goofed by defining conflicting tags in multiple places themselves.
4. `next/head` injects the results into the page.

## The benefit

Take viewport meta tags, for example. If the user wants to define the viewport themselves, they want one of two things:

1. For the viewport tag to always stay the same no matter what.
2. For the viewport tag to change based on one or more conditions.

In situation #&zwnj;1, `document/head` makes more sense. No reason to keep injecting that over and over if it isn't going to change. Conversely, in situation #&zwnj;2, `next/head` makes more sense. The only invalid situation is when a user puts viewport tags in both places, because you can't have the viewport always stay the same _and_ change sometimes. It's nearly certain that that user has made a mistake and needs to make a decision whether they want the viewport tag to change or not, so that's what they should be warned about.

## Summary

So does that sound possible, letting `next/head` know what elements were in `document/head` and adjusting its filtering so that `document/head` always wins conflicts? I think that would make the purpose of `document/head` much clearer, turning it into

- `document/head` is for things that should never ever ever change throughout your entire app.
- `next/head` is for things that you might want to change sometimes.

This is in contrast to the current `document/head` policy, which is essentially "`document/head` is for things that should never ever ever change except for the things that _could_ conflict with `next/head` _if_ you do something wrong."

If we make sure `document/head` always wins conflicts, instead of using the [current warning](https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/errors/no-document-viewport-meta.md), which kind of means "You _could_ mess things up now if you do something wrong", we could wait to warn users until they create `document/head`/`next/head` conflicts on their own, in which case the warning would be "You _did_ mess things up—you should go decide whether you want those tags to change or not."

Gosh this was far longer than I intended. 🤦🏼‍♂️

**Comment by willemmulder** [comment]
Thanks again for investigating, and writing this all down @TyMick !

Now, how would we turn this idea into actual code? Should we create an issue for it, or are Next.js devs actively monitoring these kind of ideas?

**Reply by TyMick** [reply]
Yeah, I'd originally tried to make this into an issue, but the issue template configuration they'd set up at https://github.com/vercel/next.js/issues/new/choose at the time only allowed actual Issues to be created for bug reports. Looks like their config there is a little more permissive now, and I'm also seeing a "Create issue from discussion" button in the sidebar here that says it'll carry over existing conversation, so I'll go ahead and click that button and see what happens!

**Reply by TyMick** [reply]
I take it back, the issue templates are _not_ more permissive; the [Feature Request template](https://github.com/vercel/next.js/issues/new?assignees=&labels=template%3A+story&template=3.feature_request.yml) still says it'll move the resulting issue into the GitHub Discussions "Ideas" section. But this isn't _really_ a feature request. Next.js just currently handles conflicts between `document/head` and `next/head` poorly, and it could use some improvement. I'll make an issue with the bug report template instead—this issue is closer to a bug than a feature request.

**Reply by willemmulder** [reply]
Yes I agree it's more a bug than a feature. Once you have a ticket, I'll vote there!

That said (but not relevant for this) I _do_ wonder how feature requests are monitored by Vercel once they float around in the "Ideas" section?

**Reply by TyMick** [reply]
My guess is that they only pay attention to discussions that have been upvoted some significant number of times, and since this issue is low-concept (difficult to pitch with a succinct premise), this discussion hasn't attracted many. And it doesn't help that this Discussions tab isn't very common on GitHub, so many folks probably don't realize that upvoting helps a Discussion's visibility (to the point, this Discussion has gotten more ❤️s than ↑s).

At the same time, I don't envy the Next.js team's having to deal with the staggering number of new Issues a project of this size and popularity generates. This may not be a bad system for true feature requests.

**Reply by TyMick** [reply]
Anyway, I just created a new issue at #30745! I rearranged the arguments from this discussion post to hopefully pitch it better as a real bug. 🤞🏼

**Reply by willemmulder** [reply]
Yes they have more than enough on their plate I reckon, but they probably/hopefully also have a strategy to triage and fix those smaller issues. Let's see; I +1 the issue! 👍

**Reply by willemmulder** [reply]
Thanks a lot Ty!

**Reply by TyMick** [reply]
I suspect that the only reliable way to turn it into code, though, would be to submit a PR. I'd love to at some point but don't have the time right now. And I can see this not being a priority for the core team, since while the current behavior of `document/head` is extremely confusing, the workaround (once a user finally finds it) is fairly simple: put your tag in the less optimal location of `next/head` in `_app.js` instead.

**Reply by TyMick** [reply]
Well, #30745 got converted into Discussion #32269. It was worth a try. And at least the closed and locked #30745 gets to stay in the Issues list for posterity, since I doubt many people search Discussions when they have problems. But a PR is definitely the only way this is going to get fixed.
